### PR TITLE
tools/c7n-mailer - redo c7n-mailer docker packaging for azure support

### LINKED ISSUE #3837: C7n_mailer - Docker image doesn't work with Azure
### Issue:
Out of the box, the c7n-mailer docker image doesn't seem to work with Azure. I have reasons to use the docker image versus an Azure function.

When adding in:
```
RUN cd cloud-custodian/tools/c7n_azure && \
    pip install -r ./requirements.txt && \
    pip install -e .
```
the error returned is 
```
unable to execute 'gcc': No such file or directory error: command 'gcc' failed with exit status 1
```
Upon researching, found https://github.com/docker-library/python/issues/312, but the image still does not build successfully when this is added in:
```
RUN apk add --no-cache --virtual .build-deps gcc musl-dev
RUN pip install cython
RUN apk del .build-deps gcc musl-dev
```
#### summary
I have successfully built an image using the stretch python3.7.x docker image, but it would be nice to have this work with the Alpine image, which is significantly smaller.

### PR DESCRIPTION
**Pull request by kapilt**
fixes #3837

### CODE REVIEW

**Comment by kapilt** [comment]
this has an after effect of significantly increasing the docker image size 200mb -> 1.6gb, @justinhauer was going to have a look to see if it could be slimmed.

**Comment by justinhauer** [comment]
@kapilt @logachev The following Docker image builds successfully with the 3.7.3slim-stretch image.`build-essential` was necessary the the Azure components to build successfully. Image size is down to 789mb. Still too large IMO, Alpine build would be ideal, but less than half the size of the stretch image.

#### Dockerfile:

```
FROM python:3.7.3-slim-stretch
WORKDIR /tmp
ENV PATH="/root/bin/:/tmp/:${PATH}"
COPY execution.sh ./
RUN apt-get --yes update \
 && apt-get --yes upgrade \
 && apt-get --yes install build-essential \
 && apt-get --yes install git \
 && git clone https://github.com/capitalone/cloud-custodian.git ./cloud-custodian \
 && apt-get --yes remove git \
 && pip install ./cloud-custodian \
 && pip install ./cloud-custodian/tools/c7n_mailer \
 && pip install ./cloud-custodian/tools/c7n_azure
CMD ["/bin/bash", "/tmp/execution.sh"]
```

**Comment by kapilt** [comment]
@justinhauer thanks, what's in execution.sh?

**Comment by kapilt** [comment]
I think the removes could be structured for post install .. after the pip lines, and then also target build-essential and remove /var/cache/apt

**Comment by kapilt** [comment]
I still want the out of tree builds so that this works with branches/ci without recursive checkouts etc. I did notice `.git` was missing in .dockerignore that plus excluding tests, gets a slim build down to 571mb for me with the following dockerfile

```
FROM python:3.7-slim-stretch
ADD . /src
WORKDIR /src
RUN apt-get --yes update && apt-get --yes upgrade && apt-get --yes install python-dev build-essential \
 && pip3 install -r requirements.txt -e . \
 && pip3 install -r tools/c7n_azure/requirements.txt -e tools/c7n_azure \
 && pip3 install -r tools/c7n_mailer/requirements.txt -e tools/c7n_mailer \
 && apt-get --yes remove build-essential python-dev \
 && apt-get purge --yes --auto-remove -o APT::AutoRemove::RecommendsImportant=false \
 && rm -Rf /var/cache/apt/ \
 && rm -rf /var/lib/apt/lists/*
```

**Comment by justinhauer** [comment]
@kapilt ^ that's great! I'm basically just running `c7n-mailer --config mailer.yml --run` in execution.sh, so perhaps I should be using your version where the entrypoint is `c7n-mailer`, but I just don't understand how I'd use that specific entrypoint in a k8s cronjob, hence me just running that shell script. To your point, yeah break out the tree of builds.

**Comment by kapilt** [comment]
note if you want to build locally you should use a clean checkout of custodian, as the .dockerignore won't excluded virtualenvs in the tools directory.. i don't know that we're going to get much smaller then this, the transitive azure dep stack comes in at 300mb+ afaics in site-packages by itself per `dive` inspection.

**Comment by kapilt** [comment]
re k8s cronjob I think you just drop the args, the image should default to running the ENTRYPOINT  if I'm reading the docs for it correctly https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/ although maybe I'm mixing docker and k8s worlds.

**Comment by justinhauer** [comment]
Are you saying change my entrypoint to `c7n-mailer --config mailer.yml --run` drop the `args` in the example, then I can drop the shell script all together?

**Comment by kapilt** [comment]
yeah i would think you should be able to use `args: --config mailer.yml --run`

the entrypoint is basically a default execution script... ie. with just docker its the following .. but my k8s is rusty, so ymmv.

```
docker run -vconf:/conf cloudcustodian/mailer --config conf/mailer.yml --run
```

File: .dockerignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+.git/
```
Comment (kapilt) [inline .dockerignore]: this might cause issues with some of the other docker images that are doing `-e` installs

**Comment by kapilt** [comment]
so not adding `.git` and `tests` to dockerignore increases image size backup to 737mb as the ADD/COPY layer still tracks that content even though we remove the dir later, i'm still exploring here because that precludes some use cases of using docker to run tests as well (the .git work would need to be coordinated with other dockerfiles in the repo as it precludes the use of pip install -e) , and there isn't any way to specify `.dockerignore` as a parameter to build. 

still playing around to see if i can get add/copy to do the right thing wrt to selectively copying files which would be ideal as well and could be localized to individual dockerfiles

**Comment by logachev** [comment]
@kapilt you can probably try this approach: https://docs.docker.com/engine/examples/dotnetcore/

Build & install all dependencies in some `build-env` container and copy `site-packages` (and required scripts) to the real image.

**Comment by kapilt** [comment]
yeah.. i've looked at the chained docker build stuff before, it works well for compiled language setups. in the context of python many of the installed files/libs are runtime deps, we do the one liner for install build tools/ compile/remove build tools above. The particular focus area at the moment is just getting add/copy syntax write so we can only selective move in the source bits that we care about instead of bringing in .git and tests, which we don't want to include in .dockerignore. there's all kinds of nuance about trailing slashes in copy/add src and destinations i'm still working through to see i can get the desired result, i'm hopeful just playing with it a bit more will get optimal result. going to chained build setup would also work but would basically mean copy all of /usr/local  ~440mb  most of the bloat i'm trying to cover off in is in an intermediate layer around /src/

**Comment by kapilt** [comment]
splitting up the ADD commands seems to get to the desired result, it adds a few more small layers, but the final image weighs in at 573mb which is a nice win over the current 1.6gb.

```
FROM python:3.7-slim-stretch
# Assemble c7n by dir, if anyone can combine that might be helpful
ADD setup.py README.rst requirements.txt /src/
ADD c7n /src/c7n/
ADD tools /src/tools/
WORKDIR /src
RUN apt-get --yes update && apt-get --yes upgrade && apt-get --yes install python-dev build-essential \
 && pip3 install -r requirements.txt  . \
 && pip3 install -r tools/c7n_azure/requirements.txt tools/c7n_azure \
 && pip3 install -r tools/c7n_mailer/requirements.txt tools/c7n_mailer \
 && apt-get --yes remove build-essential python-dev \
 && apt-get purge --yes --auto-remove -o APT::AutoRemove::RecommendsImportant=false \
 && rm -Rf /var/cache/apt/ \
 && rm -Rf /var/lib/apt/lists/* \
 && rm -Rf /src/

ENTRYPOINT ["c7n-mailer"]
```

**Pull request merged**